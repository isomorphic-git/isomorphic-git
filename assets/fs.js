BrowserFS.configure({ fs: 'IndexedDB', options: {} }, function (err) {
  if (err) return console.log(err)
  window.fs = BrowserFS.BFSRequire('fs')
  window.dir = 'tutorial'
})
