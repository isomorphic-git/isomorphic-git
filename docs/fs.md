---
title: fs
sidebar_label: fs
---

You need to pass a file system into `isomorphic-git` functions that do anything that involves files (which is most things in git).

In Node, you can pass the builtin `fs` module.
In the browser it's more involved because there's no standard 'fs' module.
But you can use any module that implements enough of the `fs` API.


## Node's fs

If you're only using isomorphic-git in Node, you can just use the native `fs` module:

```js
const git = require('isomorphic-git');
const fs = require('fs');
const files = await git.listFiles({ fs, dir: __dirname });
console.log(files)
```

## LightningFS

If you are writing code for the browser, you will need something that emulates the `fs` API.
While BrowserFS (see next section) has more features, [LightningFS](https://github.com/isomorphic-git/lightning-fs) might very well fit your needs.
It was designed from scratch for `isomorphic-git` (by the same author) to eek out more performance
for fewer bytes. As an added bonus it's dead simple to configure.

```html
<script src="https://unpkg.com/@isomorphic-git/lightning-fs"></script>
<script src="https://unpkg.com/isomorphic-git"></script>
<script>
const fs = new LightningFS('my-app')
const files = git.listFiles({ fs, dir: '/' });
console.log(files);
</script>
```

You can configure LightningFS to load files from an HTTP server as well, which makes it easy to prepopulate a browser file system
with a directory on your server. See the LightningFS documentation for an example of how to do this.

## BrowserFS

At the time of writing, the most complete option is [BrowserFS](https://github.com/jvilk/BrowserFS).
It has a few more steps involved to set up than in Node, as seen below:

```html
<script src="https://unpkg.com/browserfs@beta"></script>
<script src="https://unpkg.com/isomorphic-git"></script>
<script>
BrowserFS.configure({ fs: "IndexedDB", options: {} }, function (err) {
  if (err) return console.log(err);
  const fs = BrowserFS.BFSRequire("fs");
  const files = git.listFiles({ fs dir: '/' });
  console.log(files);
});
</script>
```

Besides IndexedDB, BrowserFS supports many different backends with different performance characteristics (some backends support sync operations, some only async), as well as different features such as proxying a static file server as a read-only file system, mounting ZIP files as file systems, or overlaying a writeable in-memory filesystem on top of a read-only filesystem.
You don't need to know all these features, but familiarizing yourself with the different options may be necessary if you hit a storage limit or performance bottleneck in the IndexedDB backend I suggested above.

An [advanced example usage](https://github.com/isomorphic-git/isomorphic-git/blob/53f2e909030adb1c6ae855b14f3a2474ca93ce71/__tests__/__helpers__/FixtureFS.js#L12) is in the old unit tests for isomorphic-git.
It uses HTTPRequestFS to mount (read-only) the test fixtures directory which is stored on the server, then adds a read-write InMemoryFS layer using OverlayFS so that the tests can modify files locally.
In between tests it empties the InMemoryFS, restoring the file system to a pristine state.
The current unit tests use LightningFS instead, which was built with this HTTP-backed overlay behavior by default, because I find it so useful.


# Implementing your own `fs`

There are actually TWO possible interfaces for an `fs` object: the classic "callback" API and the newer "promise" API. If your `fs` object provides an enumerable `promises` property, `isomorphic-git` will use the "promise" API _exclusively_.

## Using the "callback" API

A "callback" `fs` object must implement the following subset of node's `fs` module:

  - [fs.readFile(path[, options], callback)](https://nodejs.org/api/fs.html#fs_fs_readfile_path_options_callback)
  - [fs.writeFile(file, data[, options], callback)](https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback)
  - [fs.unlink(path, callback)](https://nodejs.org/api/fs.html#fs_fs_unlink_path_callback)
  - [fs.readdir(path[, options], callback)](https://nodejs.org/api/fs.html#fs_fs_readdir_path_options_callback)
  - [fs.mkdir(path[, mode], callback)](https://nodejs.org/api/fs.html#fs_fs_mkdir_path_mode_callback)
  - [fs.rmdir(path, callback)](https://nodejs.org/api/fs.html#fs_fs_rmdir_path_callback)
  - [fs.stat(path[, options], callback)](https://nodejs.org/api/fs.html#fs_fs_stat_path_options_callback)
  - [fs.lstat(path[, options], callback)](https://nodejs.org/api/fs.html#fs_fs_lstat_path_options_callback)
  - [fs.readlink(path[, options], callback)](https://nodejs.org/api/fs.html#fs_fs_readlink_path_options_callback) (optional [¹](#footnote-1))
  - [fs.symlink(target, path[, type], callback)](https://nodejs.org/api/fs.html#fs_fs_symlink_target_path_type_callback) (optional [¹](#footnote-1))
  - [fs.chmod(path, mode, callback)](https://nodejs.org/api/fs.html#fs_fs_chmod_path_mode_callback) (optional [²](#footnote-2))
  - [fs.rm(path[, options], callback)](https://nodejs.org/api/fs.html#fs_fs_rm_path_options_callback) (optional [³](#footnote-3))

Internally, `isomorphic-git` wraps the provided "callback" API functions using [`pify`](https://www.npmjs.com/package/pify).

As of node v12 the `fs.promises` API has been stabilized. (`lightning-fs` also provides a `fs.promises` API!) Nowadays, wrapping the callback functions
with `pify` is redundant and potentially less performant than using the native promisified versions. Plus, if you're writing your own `fs` implementation,
the `fs.promises` API lets you write straightforward implementations using `async / await` without the messy optional argument handling the callback API needs.
Therefore a second API is now supported...

## Using the "promise" API (preferred)

A "promise" `fs` object must implement the same set functions as a "callback" implementation, but it implements the promisified versions, and they should all be on a property called `promises`:

  - [fs.promises.readFile(path[, options])](https://nodejs.org/api/fs.html#fs_fspromises_readfile_path_options)
  - [fs.promises.writeFile(file, data[, options])](https://nodejs.org/api/fs.html#fs_fspromises_writefile_file_data_options)
  - [fs.promises.unlink(path)](https://nodejs.org/api/fs.html#fs_fspromises_unlink_path)
  - [fs.promises.readdir(path[, options])](https://nodejs.org/api/fs.html#fs_fspromises_readdir_path_options)
  - [fs.promises.mkdir(path[, mode])](https://nodejs.org/api/fs.html#fs_fspromises_mkdir_path_options)
  - [fs.promises.rmdir(path)](https://nodejs.org/api/fs.html#fs_fspromises_rmdir_path)
  - [fs.promises.stat(path[, options])](https://nodejs.org/api/fs.html#fs_fspromises_stat_path_options)
  - [fs.promises.lstat(path[, options])](https://nodejs.org/api/fs.html#fs_fspromises_lstat_path_options)
  - [fs.promises.readlink(path[, options])](https://nodejs.org/api/fs.html#fs_fspromises_readlink_path_options) (optional [¹](#footnote-1))
  - [fs.promises.symlink(target, path[, type])](https://nodejs.org/api/fs.html#fs_fspromises_symlink_target_path_type) (optional [¹](#footnote-1))
  - [fs.promises.chmod(path, mode)](https://nodejs.org/api/fs.html#fs_fspromises_chmod_path_mode) (optional [²](#footnote-2))
  - [fs.promises.rm(path[, options])](https://nodejs.org/api/fs.html#fs_fspromises_rm_path_options) (optional [³](#footnote-3))

---

<a id="footnote-1">¹</a> `readlink` and `symlink` are only needed to work with git repos that contain symlinks.

<a id="footnote-2">²</a> Right now, isomorphic-git rewrites the file if it needs to change its mode. In the future, if `chmod` is available it will use that.

<a id="footnote-3">³</a> Only called with `recursive: true` option. A fallback implementation is provided if not implemented.
