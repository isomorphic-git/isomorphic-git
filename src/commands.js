export * from './commands/add'
export * from './commands/branch'
export * from './commands/checkout'
export * from './commands/clone'
export * from './commands/commit'
export * from './commands/config'
export * from './commands/currentBranch'
export * from './commands/fetch'
export * from './commands/findRoot'
export * from './commands/getRemoteInfo'
export * from './commands/indexPack'
export * from './commands/init'
export * from './commands/listBranches'
export * from './commands/listFiles'
export * from './commands/listTags'
export * from './commands/log'
export * from './commands/merge'
export * from './commands/pull'
export * from './commands/push'
export * from './commands/readObject'
export * from './commands/remove'
export * from './commands/resolveRef'
export * from './commands/sign'
export * from './commands/status'
export * from './commands/verify'
export * from './commands/version'
