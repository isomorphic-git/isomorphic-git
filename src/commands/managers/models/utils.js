export { rm } from './utils/rm'
export { exists } from './utils/exists'
export {
  flatFileListToDirectoryStructure
} from './utils/flatFileListToDirectoryStructure'
export { default as fs } from './utils/fs'
export { lock, unlock } from './utils/lockfile'
export { mkdirs } from './utils/mkdirs'
export { read } from './utils/read'
export { resolveRef } from './utils/resolveRef'
export { sleep } from './utils/sleep'
export { write } from './utils/write'
export { pkg } from './utils/pkg'
