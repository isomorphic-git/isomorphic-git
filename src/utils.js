export * from './utils/auth'
export * from './utils/calculateBasicAuthHeader'
export * from './utils/calculateBasicAuthUsernamePasswordPair'
export * from './utils/comparePath'
export * from './utils/flatFileListToDirectoryStructure'
export * from './utils/log'
export * from './utils/oauth2'
export * from './utils/padHex'
export * from './utils/pkg'
export * from './utils/shasum'
export * from './utils/sleep'
export * from './utils/t'
