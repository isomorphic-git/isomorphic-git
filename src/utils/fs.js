var fs
export default function () {
  return fs
}
export function setfs (newFs) {
  fs = newFs
}
