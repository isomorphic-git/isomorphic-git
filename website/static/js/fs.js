// Initialize isomorphic-git with a file system
window.fs = new LightningFS('fs')
git.plugins.set('fs', window.fs)

// make a Promisified version for convenience
window.pfs = window.fs.promises

window.dir = '/tutorial'
