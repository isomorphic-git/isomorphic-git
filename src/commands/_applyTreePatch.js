// @ts-check
import { FileSystem } from '../models/FileSystem.js'
import { GitTree } from '../models/GitTree.js'
import { writeObject } from '../storage/writeObject.js'
import { join } from '../utils/join.js'
import { cores } from '../utils/plugins.js'

import { TREE } from './TREE.js'
import { _TREEPATCH } from './_TREEPATCH.js'
import { walkBeta1 } from './walkBeta1.js'

/**
 *
 * @typedef {import('./_diffTree').TreePatch} TreePatch
 */

/**
 *
 * @typedef {import('./readObject').TreeEntry} TreeEntry
 */

/**
 * Apply a single `TreePatch` to a tree and return a tree
 *
 * @param {Object} args
 * @param {string} [args.core = 'default'] - The plugin core identifier to use for plugin injection
 * @param {FileSystem} [args.fs] - [deprecated] The filesystem containing the git repo. Overrides the fs provided by the [plugin system](./plugin_fs.md).
 * @param {string} [args.dir] - The [working tree](dir-vs-gitdir.md) directory path
 * @param {string} [args.gitdir=join(dir,'.git')] - [required] The [git directory](dir-vs-gitdir.md) path
 * @param {string} args.base - THS SHA-1 object id of the base tree
 * @param {TreePatch} args.treePatch - A tree patch object
 *
 * @returns {Promise<string>} The SHA-1 object id of the new tree
 * @see TreePatch
 *
 */
export async function _applyTreePatch ({
  core = 'default',
  dir,
  gitdir = join(dir, '.git'),
  fs: _fs = cores.get(core).get('fs'),
  base,
  treePatch
}) {
  try {
    const fs = new FileSystem(_fs)
    const tree = TREE({ core, dir, gitdir, fs, ref: base })
    const patch = _TREEPATCH({ patch: treePatch })

    const results = await walkBeta1({
      trees: [tree, patch],

      map: async ([repo, patch]) => {
        // label entries
        await Promise.all([
          repo.populateStat(),
          patch.populateStat(),
          repo.populateHash(),
          repo.populateHash()
        ])
        // handle deletions
        if (patch.exists && patch.ops) {
          if (patch.ops.includes('rm') && !patch.ops.includes('mkdir')) {
            return
          }
          if (patch.ops.includes('rmdir') && !patch.ops.includes('write')) {
            return
          }
        }
        const entry = {
          path: repo.basename,
          mode: void 0,
          oid: void 0,
          type: void 0
        }
        if (repo.exists) {
          entry.mode = repo.mode
          entry.oid = repo.oid
          entry.type = repo.type
        }
        // handle changes
        if (patch.exists && patch.ops) {
          if (patch.ops.includes('write')) {
            entry.type = 'blob'
            entry.oid = patch.after
            if (repo.exists) {
              entry.mode = repo.mode
            } else {
              entry.mode = '100655'
            }
          }
          if (patch.ops.includes('chmod+x')) {
            entry.mode = '100755'
          } else if (patch.ops.includes('chmod-x')) {
            entry.mode = '100655'
          }
          if (patch.ops.includes('mkdir')) {
            entry.type = 'tree'
            entry.mode = '040000'
          }
        }
        return entry
      },
      /**
       * @param {TreeEntry} [parent]
       * @param {Array<TreeEntry>} children
       */
      reduce: async (parent, children) => {
        const entries = children.filter(Boolean) // remove undefineds

        // automatically delete directories if they have been emptied
        if (parent && parent.type === 'tree' && entries.length === 0) return

        if (entries.length > 0) {
          const tree = new GitTree(entries)
          const object = tree.toObject()
          const oid = await writeObject({ fs, gitdir, type: 'tree', object })
          parent.oid = oid
        }
        return parent
      }
    })
    return results.oid
  } catch (err) {
    err.caller = 'git._applyTreePatch'
    throw err
  }
}
