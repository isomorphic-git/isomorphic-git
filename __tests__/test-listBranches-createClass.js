/* global test describe expect */
import fs from 'fs'
import { createClass } from '../dist/for-node/utils'
import { listBranches } from '../dist/for-node/commands'

const Git = createClass({ listBranches })

describe('listBranches', () => {
  test('listBranches', async () => {
    let repo = new Git({
      fs,
      gitdir: '__tests__/__fixtures__/test-listBranches.git'
    })
    let commits = await repo.listBranches()
    expect(commits).toMatchSnapshot()
  })
})
