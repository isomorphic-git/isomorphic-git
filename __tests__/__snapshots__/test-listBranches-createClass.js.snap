// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listBranches listBranches 1`] = `
Array [
  "feature/supercool",
  "master",
  "test-branch",
  "greenkeeper/initial",
]
`;
